rustls/src/server/handy.rs

```diff
@@ -160,7 +160,43 @@ impl ResolvesServerCertUsingSni {
             .map_err(|_| Error::General("Bad DNS name".into()))
             .map(|dns| dns.to_owned_normalised())?;
 
-        ck.cross_check_end_entity_cert(Some(&checked_name.borrow()))?;
+        // Check the certificate chain for validity:
+        // - it should be non-empty list
+        // - the first certificate should be parsable as a x509v3,
+        // - the first certificate should quote the given server name
+        //   (if provided)
+        //
+        // These checks are not security-sensitive.  They are the
+        // *server* attempting to detect accidental misconfiguration.
+
+        // Always reject an empty certificate chain.
+        let end_entity_cert = ck.end_entity_cert().map_err(|_| {
+            Error::General("No end-entity certificate in certificate chain".to_string())
+        })?;
+
+        // Reject syntactically-invalid end-entity certificates.
+        let end_entity_cert =
+            webpki::EndEntityCert::try_from(end_entity_cert.as_ref()).map_err(|_| {
+                Error::General(
+                    "End-entity certificate in certificate \
+                                  chain is syntactically invalid"
```

Comment:
style nit: This is copied verbatim from the `sign.rs` version, but is there a reason to keep the `\` escaped newlines? It looks like both instances fit into a normal contiguous string literal without any `rustfmt` weirdness.

Reply:
have fixed that, not sure where they came from originally -- i'm not sure rustfmt splits strings like this unless the code is really wide.